typing/ctype.ml

```diff
@@ -878,15 +878,17 @@ let update_level env level ty =
       update_level env level true ty
   end
 
-(* Generalize and lower levels of contravariant branches simultaneously *)
+(* Lower level of type variables inside contravariant branches *)
 
-let rec generalize_expansive env var_level visited ty =
+let rec generalize_expansive env var_level visited contra ty =
   let ty = repr ty in
   if ty.level = generic_level || ty.level <= var_level then () else
-  if not (Hashtbl.mem visited ty.id) then begin
-    Hashtbl.add visited ty.id ();
+  if try Hashtbl.find visited ty.id < contra with Not_found -> true then begin
+    Hashtbl.add visited ty.id contra;
```

Comment:
Could I suggest something like:

```ocaml
let visit =
  match Hashtbl.find visited ty.id with
  | previous_contra -> not previous_contra && contra
  | exception Not_found -> true
in
if visit then begin
...
```
I find the current code a bit hard to read.

Reply:
It adds 4 lines, but I agree it's more readable.